### PR TITLE
Propagate ClientRequestContext to client callbacks

### LINKED ISSUE #760: ClientRequestContext is not propagated to client callbacks
`THttpClientInvocationHandler` and `THttpClientDelegate` do not seem to propagate `ClientRequestContext` properly to `AsyncMethodCallback` and the callbacks registered to `RpcResponse`.

Similarly, `HttpClientDelegate` doesn't seem to do so either.

`UserClient.execute(...)` pushes the client context, but that is not enough.

### LINKED ISSUE #1083: A user may get the wrong RequestContext when contexts are nested
When a user writes a service that invokes a request using an Armeria client, the `ServiceRequestContext` will be *hidden* by the client's `ClientRequestContext`. This can cause various unexpected hard-to-spot behaviors, especially when a user is using frameworks like RxJava, such as:

- When a user accesses the properties of `RequestContext`, a wrong value may be returned.
- When a user assumes the context is `ServiceRequestContext`, a `ClassCastException` will be raised.
- When a user access an attribute, the attribute may be missing. A user has to copy all attributes using `onChild()` to work around this.

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
`ClientRequestContext` is not propagate to `AsyncMethodCallback` and the callbacks registered to `RpcResponse`.

Modifications:
- Push `ClientRequestContext` into thread-local in `DeferredHttpResponse`, `DecodedHttpResponse` and `HttpResponseDuplicator` when invoking:
  - `Subscriber.onSubscribe()`
  - `Subscriber.onError()`
  - `Subscrbier.onComplete()`
  - Please note that the context is not pushed when calling `Subscriber.onNext()`
- `DeferredStreamMessage` calls `Subscriber.onSubscribe()` to downstream always after it subscribes to upstream

Result:
- You can now get `ClientRequestContext` in `AsyncMethodCallback` and `Subscribers` which subscribes to `HttpResponse` using `RequestContext.currentOrNull()`.
- (Breaking) `DeferredHttpResponse` is removed from public APIs. Use `HttpResponse.from()`.
- Close #760

To-do:
- Should fix #1083 before this is merged

### CODE REVIEW

**Comment by anuraaga** [comment]
We already do this for gRPC here

https://github.com/line/armeria/blob/master/grpc/src/main/java/com/linecorp/armeria/client/grpc/ArmeriaClientCall.java#L350

And I think doing something similar in thrift client makes lots of sense.

I'm less sure about the subscriber - subscriber, and even http response, are concepts that don't related to server or client. It seems pretty hacky to have them best-effort push a client context. Also why not push for `onNext`?

Is it not enough that when subscribing we can specify `ctx.contextAwareEventLoop` when subscribing?

**Comment by minwoox** [comment]
> Is it not enough that we can specify ctx.contextAwareEventLoop when subscribing?

As described in https://github.com/line/centraldogma/pull/450/, it's hard to get `ClientRequestContext` if we don't insert a decorator to store the context. So we cannot call `ctx.contextAwareEventLoop`.

I thought the user will be expecting the subscriber has the context in the thread-local. But if you think this is too much then I can just do pushing context for `AsyncMethodCallback`.
But one thing I noticed is that If we push the context and if we implement `RequestContext.parent()`, then the user does not have to use `makeContextAware` most of the time.
```java
@Override
HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
    final CompletableFuture<AggregatedHttpResponse> future = webClient.execute(req).aggregate();
    future.handle((res, cause) -> { // Do not have call ctx.makeContextAware(future)
       webClient2.execute(newReq).aggregate();
       ...
    });
    ...
}
```


For `onNext`, it can be called multiple times unlike other methods, I have chosen not to push the context. But I think if we need it, I can do it.

**Comment by anuraaga** [comment]
I've thought about automatic context propagation in the context of tracing, but run into the wall that even if Armeria handles it well, anytime using something else like a DB will require context awareness so have figured it's either all-in or all-out in some sense. But I guess outside of tracing, implicit transfer could still be good enough for many cases. For reference, `grpc-context` has no automatic transfer at all (there isn't even an automatic context, user's create the first one if they need it, for example in a tracing interceptor).

It reminds me of your question a while ago of whether we should even push context when a method signature provides access to it. `RequestContext.current()` I think has it's primary purpose as giving access to the context from non-Armeria RPC APIs where it's otherwise impossible - I don't think it's as important to push it elsewhere. For the `Subscriber`, I think semantically it is too strange to only push it for most, but not all the callbacks - but I agree the performance impact of mounting on `onNext` could be huge, and as an example, in gRPC we would have `onNext` for every binary payload in a stream but only really need to mount when notifying of a parsed message. So if all can't be context-aware, maybe that's a good enough argument for none of them to?

File: core/src/test/java/com/linecorp/armeria/client/ClientRequestContextPushedOnSubscriberTest.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+            public void onError(Throwable t) {}
+
+            @Override
+            public void onComplete() {
+                requestContexts.add(RequestContext.current());
+                latch.countDown();
+            }
+        }, nextEventLoop());
+        latch.await();
+        assertThat(requestContexts.size()).isEqualTo(2);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/ClientRequestContextPushedOnSubscriberTest.java]: This assertion is unnecessary.

File: core/src/test/java/com/linecorp/armeria/client/ClientRequestContextPushedOnSubscriberTest.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+                requestContexts.add(RequestContext.current());
+                latch.countDown();
+            }
+
+            @Override
+            public void onComplete() {}
+        }, nextEventLoop());
+
+        latch.await();
+        assertThat(requestContexts.size()).isEqualTo(2);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/ClientRequestContextPushedOnSubscriberTest.java]: Ditto

File: core/src/test/java/com/linecorp/armeria/client/ClientRequestContextPushedOnSubscriberTest.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+                latch.countDown();
+            }
+
+            @Override
+            public void onComplete() {}
+        }, nextEventLoop());
+        res.abort();
+
+        latch.await();
+        assertThat(requestContexts.size()).isEqualTo(2);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/ClientRequestContextPushedOnSubscriberTest.java]: Ditto

File: core/src/test/java/com/linecorp/armeria/client/ClientRequestContextPushedOnSubscriberTest.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                requestContexts.add(RequestContext.current());
+                latch.countDown();
+            }
+
+            @Override
+            public void onComplete() {}
+        }, nextEventLoop());
+
+        latch.await();
+        assertThat(requestContexts.size()).isEqualTo(2);
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/client/ClientRequestContextPushedOnSubscriberTest.java]: Ditto

File: thrift/src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java [hunk 5]
```diff
@@ -108,7 +109,9 @@ public RpcResponse execute(ClientRequestContext ctx, RpcRequest call) throws Exc
             }
         } catch (Throwable cause) {
             reply.completeExceptionally(cause);
-            return reply;
+            try (SafeCloseable ignored = ctx.push()) {
+                return reply;
```
Comment (trustin) [inline thrift/src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java]: Why do we need to push the context here?

File: core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessage.java [hunk 6]
```diff
@@ -186,6 +182,15 @@ protected EventExecutor defaultSubscriberExecutor() {
      */
     abstract void notifySubscriberOfCloseEvent(SubscriptionImpl subscription, CloseEvent event);
 
+    /**
+     * Pushes the context to the thread-local stack if this {@link StreamMessage} has it.
+     * This method is invoked before {@link Subscriber#onSubscribe(Subscription)},
+     * {@link Subscriber#onComplete()} and {@link Subscriber#onError(Throwable)} are called.
+     */
+    protected SafeCloseable pushContextIfExist() {
+        return () -> { /* no-op */};
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessage.java]: nit: Missing space between `*/` and `}`

File: core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessageDuplicator.java [hunk 7]
```diff
@@ -171,6 +171,26 @@ public void abort(Throwable cause) {
...
+
+    /**
+     * Pushes the context to the thread-local stack if this {@link AbstractStreamMessageDuplicator} has it.
+     * This method is invoked before {@link Subscriber#onSubscribe(Subscription)},
+     * {@link Subscriber#onComplete()} and {@link Subscriber#onError(Throwable)} are called.
+     *
+     * @see #onSubscribeCalled()
+     */
+    protected SafeCloseable pushContextIfExist() {
+        return () -> { /* no-op */};
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessageDuplicator.java]: Ditto

**Comment by trustin** [comment]
> We already do this for gRPC here https://github.com/line/armeria/blob/master/grpc/src/main/java/com/linecorp/armeria/client/grpc/ArmeriaClientCall.java#L350

Oh, shouldn't we only do that in `onReady` and `onClose()` (and maybe in `onMetadata()`)? It may be premature optimization though.

@minwoox Would you mind measuring the performance impact of pushing context on `onNext()`?

**Comment by anuraaga** [comment]
`onMessage` is what invokes the gRPC stub code and while it is possible to deal with `onReady` and store context somewhere, most code doesn't even handle `onReady` - the standard callbacks are `onNext`, `onError`, and `onCompleted` for a gRPC client stub and if we didn't mount here then `onNext` would not have any access to context.

**Comment by minwoox** [comment]
I am on a training session so I might be late to reply. Sorry about that.
> Would you mind measuring the performance impact of pushing context on onNext()?

Yes, I can try. But before that, I wanted to address that the logic in the hook which is attached using `ReqeustContxte.onEnter()` is the most affected factor.
And I think I want to remove the `onEnter()` because it has two drawbacks:
- `onEnter` isn't invoked when the context is first pushed into the thread-local
  - A user attaches the hook to the context which is already pushed.
- Most of the users do not know when `onEnter` is called.
  - They do not expect the `onEnter` is called when `RequestLogListener` is invoked.
Because of these reasons, I really get rid of the API. Or at least, I want it to change to be called when `makeContextAware` series.
Let's discuss more about it on #2232 

> I've thought about automatic context propagation in the context of tracing, but run into the wall that even if Armeria handles it well, anytime using something else like a DB will require context awareness so have figured it's either all-in or all-out in some sense. But I guess outside of tracing, implicit transfer could still be good enough for many cases. For reference, grpc-context has no automatic transfer at all (there isn't even an automatic context, user's create the first one if they need it, for example in a tracing interceptor).

Most of the questions about `makeContextAware` that I got from the users are the cases that I shared in https://github.com/line/armeria/pull/2311#issuecomment-564392082. So I think if we support the automatic propagating that will really reduce the burden of the entry-level users.
But yes, they do have to know how it works to get using Armeria better.
And I also don't disagree that the explicit could be better.
So let's put it apart and focus on whether we need the context in the subscriber methods or not.
Do we agree that we need to push the context if it is not degrading the performance a lot?

**Comment by anuraaga** [comment]
I'd be fine with pushing the context in `Subscriber` if it applies to all the callbacks consistently without significant performance impact. If we find an impact, I don't think I'd go for one that only applies to some callbacks as it's a confusing restriction. Also, not necessarily in this PR, but if we do push for client, can we also push for server for consistency?

**Comment by trustin** [comment]
I'm OK with the removal of `onEnter` events, but I'm curious if there is any concrete use case we can't solve without using `onEnter()`. Thoughts, @anuraaga?

**Comment by minwoox** [comment]
Let me do the benchmark first. Thanks! 😃

**Comment by minwoox** [comment]
I realized that pushing the context before `Subscriber.onComplete()` is not enough because the callback cannot get the context if it's attached after the future is complete:
```java
CompletableFuture future = webClient.execute(req).aggregate();
future.hande((msg, cause) -> {
    // This thread will have the ClientRequestContext in the thread-local
    // most of the time.
    ...
);
....
future.hande((msg, cause) -> {
    // This thread will not have the ClientRequestContext in the thread-local
    // most of the time because the future is already completed. 
    ...
);
```

I think returning `RequestContextAwareCompletableFuture` is the only way to solve this issue. If we do that, we might not need to push the context before `Subscriber.onX`. In order to do that, we need to fix #1739 first so that we can make the `RequestContextAwareCompletableFuture` when `StreamMessage.subscribe()` is called.

I want to work on #1083 first it must be addressed before we pushing the context into the thread which executes the callbacks, and then I will do #1739.

So let me close this issue for now and reopen this if we still need to push the context before `Subscriber.onX`.
Thanks!

**Comment by anuraaga** [comment]
Thanks for investigating! By the way, what do you think about pushing the context in-band for thrift, similar to what we do in gRPC. I guess around here

https://github.com/line/armeria/blob/master/thrift/src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java#L281

**Comment by minwoox** [comment]
Yes, I have thought about it, but I thought I need to make the `DefaultRpcResponse` extends `RequestContextAwareCompletableFuture` as well so that the callback attached after the `RpcResponse` is complete will have the context. 🤔

**Comment by anuraaga** [comment]
Ah got it that probably makes more sense

File: thrift/src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java (hunk 5 above)
Comment (minwoox) [reply thrift/src/main/java/com/linecorp/armeria/client/thrift/THttpClientDelegate.java]: I don't know. 😅

**Comment by minwoox** [comment]
> I'm OK with the removal of onEnter events, but I'm curious if there is any concrete use case we can't solve without using onEnter()

@delegacy might have an idea on this because I've seen him using it.

**Pull request closed without merging**